strapi/strapi
# optional deterministic api-token generation

### What does it do?

This PR adds functionality to create deterministic api tokens.

### Why is it needed?

This is necessary in many architectures where keys need to be shared at build time prior to deployment.

This is currently not possible because api tokens are generated randomly, so even during the app bootstrap step it is not possible to create a deterministic token that can be shared to dependent services in advance.

This leads to serious issues in orchestrated environments that rely on CI / CD where strapi becomes a single point of manual deployment and potentially causes application downtime on deploy or secondary deployments post key generation in production.

### How to test it?

```javascript
const apiToken = await strapi.admin.services['api-token'].create({
  name: 'My API Token',
  description: 'This is a test API token v1.',
  type: 'full-access', // Or 'read-only', depending on the level of access you need.
  accessKey: 'b23fb4aaaba2276b1aa057cefa12e01b9ef112bb1039deb0942f317737b4f1e253c7fe716658984448b4c00ebfbb47071ba2234f6a08ff9a94e2b03914ce1e0a',
  permissions: [
              // Define specific permissions here if needed.
  ],
})

console.log(apiToken)
```

This should result in a static key generation that can be done during development, checked into secret management, and deterministically recreated in production assuming you used the same salt.
